referenceframe/urdf.go

```diff
@@ -0,0 +1,299 @@
+
+		// Set up the child link mentioned in this joint; fill out the details in the link parsing section later
+		childLink := JsonLink{ID: jointElem.Child.Link, Parent: jointElem.Name}
+
+		switch jointElem.Type {
+		case "continuous", "revolute", "prismatic":
+			// Parse important details about each joint, including axes and limits
+			jointAxes := convStringAttrToFloats(jointElem.Axis.XYZ)
+			thisJoint := JsonJoint{
+				ID:     jointElem.Name,
+				Type:   jointElem.Type,
+				Parent: jointElem.Parent.Link,
+				Axis:   spatial.AxisConfig{jointAxes[0], jointAxes[1], jointAxes[2]},
+			}
+
+			// Slightly different limits handling for continuous, revolute, and prismatic joints
+			if jointElem.Type == "continuous" {
+				thisJoint.Type = "revolute" // Currently, we treate a continuous joint as a special case of the revolute joint
+				thisJoint.Min, thisJoint.Max = utils.RadToDeg(-math.Pi), utils.RadToDeg(math.Pi)
+			} else if jointElem.Type == "prismatic" {
+				thisJoint.Min, thisJoint.Max = jointElem.Limit.Lower*1000, jointElem.Limit.Upper*1000 // from meters to mm
+			} else {
+				thisJoint.Min, thisJoint.Max = utils.RadToDeg(jointElem.Limit.Lower), utils.RadToDeg(jointElem.Limit.Upper)
+			}
+
+			mc.Joints = append(mc.Joints, thisJoint)
+
+			// Generate child link translation and orientation data, which is held by this joint per the URDF design
+			childXYZ := convStringAttrToFloats(jointElem.Origin.XYZ)
+			childRPY := convStringAttrToFloats(jointElem.Origin.RPY)
+			childEA := spatial.EulerAngles{Roll: childRPY[0], Pitch: childRPY[1], Yaw: childRPY[2]}
+			childOrient, err := spatial.NewOrientationConfig(childEA.EulerAngles())
+
+			// Note the conversion from meters to mm
+			childLink.Translation = spatial.TranslationConfig{childXYZ[0] * 1000, childXYZ[1] * 1000, childXYZ[2] * 1000}
+			childLink.Orientation = *childOrient
+
+			if err != nil {
+				return nil, err
+			}
+		case "fixed":
+			// Handle fixed joint -> static link conversion instead of adding to Joints[]
+			thisLink := JsonLink{ID: jointElem.Name, Parent: jointElem.Parent.Link}
+
+			linkXYZ := convStringAttrToFloats(jointElem.Origin.XYZ)
+			linkRPY := convStringAttrToFloats(jointElem.Origin.RPY)
+			linkEA := spatial.EulerAngles{Roll: linkRPY[0], Pitch: linkRPY[1], Yaw: linkRPY[2]}
+			linkOrient, err := spatial.NewOrientationConfig(linkEA.AxisAngles())
+
+			// Note the conversion from meters to mm
+			thisLink.Translation = spatial.TranslationConfig{linkXYZ[0] * 1000, linkXYZ[1] * 1000, linkXYZ[2] * 1000}
+			thisLink.Orientation = *linkOrient
+
+			if err != nil {
+				return nil, err
+			}
+
+			mc.Links = append(mc.Links, thisLink)
+		default:
+			return nil, NewUnsupportedJointTypeError(jointElem.Type)
+		}
+
+		if err != nil {
+			return nil, err
+		}
+
+		mc.Links = append(mc.Links, childLink)
+	}
+
+	// Handle links
+	for _, linkElem := range urdf.Links {
+		// Skip any world links
+		if linkElem.Name == "world" {
```

Comment:
q: in URDF, is it guaranteed that something named "world" has size zero?

If not, we should check, and if it's non-zero, or has an orientation, we should error.

Reply:
That's a tough question. I've never personally seen a modified world link. Note that links also do encode their own origin information (just offset data for visual, collision, or inertial elements, which, again, I have never seen).

Reply:
Follow up question, won't this allow multiple things to be named "world"?